File: mari-server/src/utils/batchSeal.js
// src/utils/batchSeal.js
// Batch payment seal generation and validation
const crypto = require('crypto');

/**
 * Generate batch seal from list of (id, amount) pairs
 * Compatible with Android BatchSeal.kt implementation
 * @param {Array<{id: string, amount: number}>} items
 * @returns {string} SHA-256 hex hash
 */
function generateBatchSeal(items) {
  const parts = items.map(({ id, amount }) => {
    // Format amount: whole numbers without decimals, floats with decimals
    const amtStr = Number.isInteger(amount) ? amount.toString() : amount.toString();
    return id + amtStr;
  }).sort(); // Lexicographic sort for deterministic order
  
  const concat = parts.join('');
  return crypto.createHash('sha256').update(concat, 'utf8').digest('hex');
}

/**
 * Verify a batch seal matches the provided items
 * @param {string} seal - Expected batch seal (hex string)
 * @param {Array<{id: string, amount: number}>} items - Batch items
 * @returns {boolean}
 */
function verifyBatchSeal(seal, items) {
  const computed = generateBatchSeal(items);
  return computed === seal;
}

module.exports = {
  generateBatchSeal,
  verifyBatchSeal
};

File: mari-app/protocol/BatchSeal.kt
package com.Mari.mobileapp.protocol

import java.security.MessageDigest

object BatchSeal {
    /**
     * Generate batch seal compatible with @Mari/shared-libs
     * Concatenate sorted (id + amountString), where amountString is without decimals for whole numbers.
     * Then SHA-256 hex of the concatenated string.
     */
    fun generate(items: List<Pair<String, Double>>): String {
        val parts = items.map { (id, amt) ->
            val amtStr = if (amt == Math.floor(amt)) {
                amt.toLong().toString()
            } else {
                amt.toString()
            }
            id + amtStr
        }.sorted()
        val concat = parts.joinToString("")
        val md = MessageDigest.getInstance("SHA-256")
        val bytes = md.digest(concat.toByteArray(Charsets.UTF_8))
        return bytes.joinToString("") { b -> "%02x".format(b) }
    }
}
